types/wlr_foreign_toplevel_management_v1.c

```diff
@@ -0,0 +1,583 @@
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &toplevel->resources) {
+		zwlr_foreign_toplevel_handle_v1_send_done(resource);
+	}
+
+	toplevel->idle_source = NULL;
+}
+
+static void toplevel_update_idle_source(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel) {
+	if (toplevel->idle_source)
+		return;
+
+	toplevel->idle_source = wl_event_loop_add_idle(toplevel->manager->event_loop,
+		toplevel_idle_send_done, toplevel);
+}
+
+void wlr_foreign_toplevel_handle_v1_set_title(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel, const char *title) {
+	if (toplevel->title) {
+		free(toplevel->title);
+	}
+	toplevel->title = strdup(title);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &toplevel->resources) {
+		zwlr_foreign_toplevel_handle_v1_send_title(resource, title);
+	}
+
+	toplevel_update_idle_source(toplevel);
+}
+
+void wlr_foreign_toplevel_handle_v1_set_app_id(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel, const char *app_id) {
+	if (toplevel->app_id) {
+		free(toplevel->app_id);
+	}
+	toplevel->app_id = strdup(app_id);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &toplevel->resources) {
+		zwlr_foreign_toplevel_handle_v1_send_app_id(resource, app_id);
+	}
+
+	toplevel_update_idle_source(toplevel);
+}
+
+static void send_output_to_resource(struct wl_resource *resource,
+		struct wlr_output *output, bool enter) {
+	struct wl_client *client = wl_resource_get_client(resource);
+	struct wl_resource *output_resource;
+
+	wl_resource_for_each(output_resource, &output->resources) {
+		if (wl_resource_get_client(output_resource) == client) {
+			if (enter) {
+				zwlr_foreign_toplevel_handle_v1_send_output_enter(resource,
+					output_resource);
+			} else {
+				zwlr_foreign_toplevel_handle_v1_send_output_leave(resource,
+					output_resource);
+			}
+		}
+	}
+}
+
+static void toplevel_send_output(struct wlr_foreign_toplevel_handle_v1 *toplevel,
+		struct wlr_output *output, bool enter) {
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &toplevel->resources) {
+		send_output_to_resource(resource, output, enter);
+	}
+
+	toplevel_update_idle_source(toplevel);
+}
+
+static void toplevel_handle_output_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_foreign_toplevel_handle_output *toplevel_output =
+		wl_container_of(listener, toplevel_output, output_destroy);
+	wlr_foreign_toplevel_handle_v1_output_leave(toplevel_output->toplevel,
+		toplevel_output->output);
+}
+
+void wlr_foreign_toplevel_handle_v1_output_enter(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel,
+		struct wlr_output *output) {
+	struct wlr_foreign_toplevel_handle_output *toplevel_output;
+	wl_list_for_each(toplevel_output, &toplevel->outputs, link) {
+		if (toplevel_output->output == output) {
+			return; // we have already sent output_enter event
+		}
+	}
+
+	toplevel_output =
+	    calloc(1, sizeof(struct wlr_foreign_toplevel_handle_output));
```

Comment:
This is missing a NULL check

Reply:
do we again `post_no_memory` to the client or just silently fail?

Reply:
For allocations not caused by client requests, we can just log an error.